File: 351002/Khmel/Task340/Publisher/Services/WriterService.cs
public class WriterService : IWriterService
{
    private readonly IRepository<Writer> _repository;
    private readonly ICacheService _cache;

    public WriterService(IRepository<Writer> repository, ICacheService cache)
    {
        _repository = repository;
        _cache = cache;
    }

    public WriterResponseTo GetById(long id)
    {
        var cacheKey = $"writer:{id}";
        var cached = _cache.GetObjectAsync<WriterResponseTo>(cacheKey).GetAwaiter().GetResult();
        
        if (cached != null)
            return cached;

        var writer = _repository.GetById(id);
        if (writer == null)
            throw new KeyNotFoundException($"Writer с id={id} не найден");

        var response = ToResponse(writer);
        _cache.SetObjectAsync(cacheKey, response, TimeSpan.FromMinutes(5)).GetAwaiter().GetResult();
        
        return response;
    }

    public IEnumerable<WriterResponseTo> GetAll()
    {
        return _repository.GetAll().Select(ToResponse);
    }

    public WriterResponseTo Create(WriterRequestTo request)
    {
        var existing = _repository.GetAll()
            .FirstOrDefault(w => w.Login == request.Login);

        if (existing != null)
            throw new ForbiddenException($"Writer с login={request.Login} уже существует");

        var writer = ToModel(request);
        var created = _repository.Create(writer);
        return ToResponse(created);
    }

    public WriterResponseTo Update(WriterRequestTo request)
    {
        var writer = ToModel(request);
        var updated = _repository.Update(writer);
        var response = ToResponse(updated);

        _cache.RemoveAsync($"writer:{request.Id}").GetAwaiter().GetResult();
        
        return response;
    }

    public void Delete(long id)
    {
        if (!_repository.Delete(id))
            throw new KeyNotFoundException($"Writer с id={id} не найден");
        
        _cache.RemoveAsync($"writer:{id}").GetAwaiter().GetResult();
    }

    private Writer ToModel(WriterRequestTo dto) => new Writer
    {
        Id = dto.Id,
        Login = dto.Login,
        Password = dto.Password,
        Firstname = dto.Firstname,
        Lastname = dto.Lastname
    };

    private WriterResponseTo ToResponse(Writer model) => new WriterResponseTo
    {
        Id = model.Id,
        Login = model.Login,
        Firstname = model.Firstname,
        Lastname = model.Lastname
    };
}

File: 351003/Suchok/publisher/src/main/java/by/liza/app/service/WriterService.java
package by.liza.app.service;

import by.liza.app.dto.request.WriterRequestTo;
import by.liza.app.dto.response.WriterResponseTo;
import by.liza.app.exception.DuplicateEntityException;
import by.liza.app.exception.EntityNotFoundException;
import by.liza.app.mapper.WriterMapper;
import by.liza.app.model.Writer;
import by.liza.app.repository.ArticleRepository;
import by.liza.app.repository.MarkRepository;
import by.liza.app.repository.WriterRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.CachePut;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.cache.annotation.Caching;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class WriterService {

    private final WriterRepository  writerRepository;
    private final ArticleRepository articleRepository;
    private final MarkRepository    markRepository;
    private final WriterMapper      writerMapper;
    private final PasswordEncoder   passwordEncoder;

    @Transactional
    @Caching(
            put  = @CachePut(value = "writers", key = "#result.id"),
            evict = @CacheEvict(value = "writers_all", allEntries = true)
    )
    public WriterResponseTo create(WriterRequestTo requestTo) {
        if (writerRepository.existsByLogin(requestTo.getLogin())) {
            throw new DuplicateEntityException(
                    "Writer with login '" + requestTo.getLogin() + "' already exists", 40301);
        }
        Writer writer = writerMapper.toEntity(requestTo);
        writer.setPassword(passwordEncoder.encode(requestTo.getPassword()));
        writer.setRole(requestTo.getRole() != null
                ? Writer.Role.valueOf(requestTo.getRole().toUpperCase())
                : Writer.Role.CUSTOMER);
        return writerMapper.toResponse(writerRepository.save(writer));
    }

    @Cacheable(value = "writers", key = "#id")
    public WriterResponseTo getById(Long id) {
        return writerMapper.toResponse(findById(id));
    }

    @Cacheable(value = "writers_all")
    public List<WriterResponseTo> getAll() {
        return writerMapper.toResponseList(writerRepository.findAll());
    }

    @Transactional
    @Caching(
            put  = @CachePut(value = "writers", key = "#result.id"),
            evict = @CacheEvict(value = "writers_all", allEntries = true)
    )
    public WriterResponseTo update(WriterRequestTo requestTo) {
        if (requestTo.getId() == null) {
            throw new EntityNotFoundException("Writer id must be provided for update", 40001);
        }
        Writer existing = findById(requestTo.getId());
        if (writerRepository.existsByLoginAndIdNot(requestTo.getLogin(), requestTo.getId())) {
            throw new DuplicateEntityException(
                    "Writer with login '" + requestTo.getLogin() + "' already exists", 40301);
        }
        writerMapper.updateEntityFromRequest(requestTo, existing);
        existing.setPassword(passwordEncoder.encode(requestTo.getPassword()));
        if (requestTo.getRole() != null) {
            existing.setRole(Writer.Role.valueOf(requestTo.getRole().toUpperCase()));
        }
        return writerMapper.toResponse(writerRepository.save(existing));
    }

    @Transactional
    @Caching(evict = {
            @CacheEvict(value = "writers",     key = "#id"),
            @CacheEvict(value = "writers_all", allEntries = true)
    })
    public void delete(Long id) {
        if (!writerRepository.existsById(id)) {
            throw new EntityNotFoundException(
                    "Writer with id " + id + " not found", 40401);
        }
        writerRepository.deleteById(id);
    }

    private Writer findById(Long id) {
        return writerRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException(
                        "Writer with id " + id + " not found", 40401));
    }
}
